src/Microsoft.Health.Fhir.CosmosDb/Features/Search/FhirCosmosSearchService.cs

```diff
@@ -138,10 +203,156 @@ public class FhirCosmosSearchService : SearchService
+            {
+                return (Array.Empty<FhirCosmosResourceWrapper>(), false);
+            }
+
+            var includes = new List<FhirCosmosResourceWrapper>();
+            bool includesTruncated = false;
+
+            if (includeExpressions.Count > 0)
+            {
+                // fetch in the resources to include from _include parameters.
+
+                var referencesToInclude = matches.SelectMany(m => m.ReferencesToInclude).Distinct().ToList();
+
+                if (referencesToInclude.Count > 0)
+                {
+                    // construct the expression typeAndId = <Include1Type, Include1Id> OR  typeAndId = <Include2Type, Include2Id> OR ...
+
+                    SearchParameterExpression expression = Expression.SearchParameter(
+                        _typeIdCompositeSearchParameter,
+                        Expression.Or(referencesToInclude.Select(r =>
+                            Expression.And(
+                                Expression.Equals(FieldName.TokenCode, 0, r.ResourceTypeName),
+                                Expression.Equals(FieldName.TokenCode, 1, r.ResourceId))).ToList()));
+
+                    var includeSearchOptions = new SearchOptions
+                    {
+                        Expression = expression,
+                        MaxItemCount = maxIncludeCount,
+                        Sort = Array.Empty<(SearchParameterInfo searchParameterInfo, SortOrder sortOrder)>(),
+                    };
+
+                    QueryDefinition includeQuery = _queryBuilder.BuildSqlQuerySpec(includeSearchOptions, Array.Empty<IncludeExpression>());
+
+                    FeedResponse<FhirCosmosResourceWrapper> includeResponse = null;
+                    do
+                    {
+                        if (includes.Count >= maxIncludeCount)
+                        {
+                            includesTruncated = true;
+                            break;
+                        }
+
+                        includeResponse = await ExecuteSearchAsync<FhirCosmosResourceWrapper>(
+                            includeQuery,
+                            includeSearchOptions,
+                            includeResponse?.ContinuationToken,
+                            cancellationToken);
+                        includes.AddRange(includeResponse);
+                    }
+                    while (!string.IsNullOrEmpty(includeResponse.ContinuationToken));
+                }
+            }
+
+            if (revIncludeExpressions.Count > 0 && !includesTruncated)
+            {
+                // fetch in the resources to include from _revinclude parameters.
+
+                foreach (IncludeExpression revIncludeExpression in revIncludeExpressions)
+                {
+                    // construct the expression resourceType = <SourceResourceType> AND (referenceSearchParameter = <MatchResourceType1, MatchResourceId1> OR referenceSearchParameter = <MatchResourceType2, MatchResourceId2> OR ...)
```

Comment:
Is there some max number of conditions we need to check for?

Reply:
Good point. Query size is [limited to 256KB](https://docs.microsoft.com/en-us/azure/cosmos-db/concepts-limits#sql-query-limits). We may need to partition in the case of forward include, where there are a lot of references.

Reply:
For _revinclude, with the page size limit we have, we should be fine.

Reply:
Added partitioning